bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_rtensor.py

```diff
@@ -9,10 +9,83 @@
 ################################################################################
 
 from .. import pythonization
-from .._rvec import _array_interface_dtype_map
+from .._rvec import _array_interface_dtype_map, _get_cpp_type_from_numpy_type
 from libROOTPythonizations import GetEndianess
 import cppyy
 
+
+def _AsRTensor(arr):
+    """
+    Adopt memory of a Python object with array interface using an RTensor.
+
+    \param[in] self Always null, since this is a module function.
+    \param[in] obj PyObject with array interface
+
+    This function returns an RTensor which adopts the memory of the given
+    PyObject. The RTensor takes the data pointer and the shape from the array
+    interface dictionary.
+    """
+    import ROOT
+    import cppyy
+    import math
+    import platform
+
+    # Get array interface of object
+    interface = arr.__array_interface__
+
+    # Get the data-pointer
+    data = interface["data"][0]
+
+    # Get the size of the contiguous memory
+    shape = interface["shape"]
+    size = math.prod(shape) if len(shape) > 0 else 0
+
+    # Get the typestring and properties thereof
+    typestr = interface["typestr"]
+    if len(typestr) != 3:
+        raise RuntimeError(
+            "Object not convertible: __array_interface__['typestr'] returned '"
+            + typestr
+            + "' with invalid length unequal 3."
+        )
+
+    if typestr[:1] != GetEndianess():
+        raise RuntimeError(
+            "Object not convertible: Endianess of __array_interface__['typestr'] does not match endianess of ROOT."
+        )
+
+    dtype = typestr[1:]
+    dtypesize = int(typestr[-1])
+    cppdtype = _get_cpp_type_from_numpy_type(dtype)
+
+    # Get strides
+    strides = arr.strides
+
+    # Infer memory layout from strides
+    memory_layout = "RowMajor"
+    if len(strides) > 1 and strides[0] < strides[-1]:
+        memory_layout = "ColumnMajor"
+
+    # Construct an RTensor of the correct data-type
+    prefix = "0x" if platform.system() == "Windows" else ""
+    klassname = f"TMVA::Experimental::RTensor<{cppdtype},std::vector<{cppdtype}>>"
+    address = ROOT.gInterpreter.Calc(
+        f"new {klassname}(reinterpret_cast<{cppdtype} *>({prefix}{data}),"
+        "{" + ",".join([f"{s}" for s in shape]) + "}, "
+        "{" + ",".join([f"{s // dtypesize}" for s in strides]) + "}, "
+        "TMVA::Experimental::MemoryLayout::" + memory_layout + ")"
+    )
+
+    # Bind the object to a Python-side proxy, taking ownership of the underlying C++ object
+    out = cppyy.bind_object(address, klassname)
```

Comment:
Can we make these lines even more Pythonic? Something like the following (untested pseudocode)
```suggestion
    tensor = TMVA.RTensor[cppdtype, std.vector[cppdtype]](
        cppyy.ll.reinterpret_cast[f"{cppdtype} *"](f"{prefix}{data}"),
        shapetuple,
        memory_layout
    )
```

Then I guess also the ownership will be automatically handled if this works

Reply:
Thanks a lot, very good suggestions! I implemented them. Let's see what the CI says now.

Unfortunately, I still have a Windows failure that I still don't understand. Maybe it will be magically gone after implementing your suggestions :)